### PR TITLE
Small code cleanup and new setting.

### PR DESCRIPTION
**Pull request by ChemiHatake**
Implemented a new setting (RecommendMods)
if its on, it'll just recommend songs normally.
but if its set to false, 
it won't recommend songs with mods (instead of using '!r nomod' every time).

Issue: #33

FINAL UPDATE:
A new translation (IllegalModelMix) will need to be translated!

### CODE REVIEW

File: tillerinobot/src/main/java/tillerino/tillerinobot/handlers/OptionsHandler.java [hunk 1]
```diff
@@ -73,7 +73,13 @@ public int compare(LanguageIdentifier o1, LanguageIdentifier o2) {
 			} else {
 				ircUser.message("Welcome Message: " + (userData.isShowWelcomeMessage() ? "ON" : "OFF"));
 			}
-		} else {
+		} else if (getLevenshteinDistance(option, "recommendmods") <= 1) {
```
Comment (Tillerino) [inline tillerinobot/src/main/java/tillerino/tillerinobot/handlers/OptionsHandler.java]: I feel like this command should somehow communicate that it only applies as a default, i.e. when no more arguments are given.

**Comment by Tillerino** [comment]
@omkelderman is this what you were thinking about regarding #33?

**Comment by omkelderman** [comment]
My initial thought when that issue arrived a while back was to have a setting where you can specify a "default set of arguments" when nothing is specified to `!r`
Like you could also use it if someone for some reason only wants HR recomendations or whatever.
So you'd have a setting `!set ??? hr` (im horrible at names) and every `!r` without args would translate to `!r hr`. In the same way you could do `!set ??? nomod` and `!r` would translate to `!r nomod`.

But this might be overkill since its probably a use case nobody will use. So I'm also ok with this solution.

***But*** I have one concern with this solution: that is that this setting is "recommendmods" where if it is true it will do the behavior the bot does now, and if it is false it will only do nomod-recomendations. Somehow I have a feeling that introducing this change will result in everyone giving a default value for this setting of false, and thus changing the default behavior for everyone! Correct me if I'm wrong, but I have a feeling that would be ***very*** bad.

Besides that I miss an option to ask the bot for recommendations *with* mods (as in, the current default behavior) *without* having to change the settings. For when I have this setting set, but want a couple recommendations with the current behavior.

**Comment by Tillerino** [comment]
I agree, an option where you can just set default options would be more flexible and more transparent (like in ppaddict). Bypass the default settings with `!r any` or something.

**Comment by omkelderman** [comment]
Ok, I just did a quick check, and my concern with the default value for the setting is invalid, turns out I completely didnt saw the `= true` in `boolean recommendModdedMaps = true;` so scratch that part of my commend, the rest still stands though.

**Comment by ChemiHatake** [comment]
@Tillerino Aight, I'll do that.

**Comment by ChemiHatake** [comment]
@Tillerino @omkelderman 
Done!

**Comment by ChemiHatake** [comment]
It loaded the other classes before loading the new enum I've made, thats why.

**Comment by omkelderman** [comment]
just random question: why have all the fancy parsing stuff in the options-handler, when you can just save it as a string, and feed that back into the RecommendationsManager if the setting has a value. I mean, RecommendationsManager has all the options-parsing logic

**Comment by ChemiHatake** [comment]
damnit there's a bug that makes the build fail

**Comment by omkelderman** [comment]
> there's a bug

Thats a `findbugs` thing (a tool which checks your code). It doesnt want you to return the internal array-object, since normally that's not what you want: you're exposing internal objects, eg you can do `arr[0] = obj` with the return-array and the original array inside the object you you got it from is changed.
Maybe that's not a problem here, idk, solution would be to clone inside the getter (which in this case is created by lombok, so you'd have to create one yourself), or add a fancy ignore annotation which probably exists.

But beside that, like i said earlier, if I where to implement this: I would just store it as a string and not parse anything at all. You only need the string with stored options for later anyway which you can feed into the RecommendationsManager which already does all the parsing.

**Comment by ChemiHatake** [comment]
@omkelderman you're right 
Ill work on it

**Comment by Tillerino** [comment]
If I'm not mistaken, `tillerino.tillerinobot.RecommendationsManager.parseSamplerSettings(OsuApiUser, String, Language)` does exactly what you want. It doesn't give very good error messages, for example if you go `relax dt` it will just throw a generic exception. But I'd rather you improve _that_ logic instead of duplicating a lot of code.

**Comment by ChemiHatake** [comment]
@Tillerino 
But we don't want users to set the RecommendMods value to DT Relax etc..,
I'm pretty sure I'd have to make a whole new method for the parsing.

**Comment by omkelderman** [comment]
> it doesn't return anything.

```
public Settings parseSamplerSettings(OsuApiUser apiUser, @Nonnull String message,
			Language lang) throws UserException, SQLException, IOException
```

Looks like a return-value to me ^^

But to give a more meaningfull answer: If the method returns you can be sure the syntax was correct, since it'll throw if it isnt. Which is good, since it throws `UserException`s which automatically gets converted into error messages show to the user.

And now im thinking about it: that `Settings` (`tillerino.tillerinobot.RecommendationsManager.Sampler.Settings`) contains pretty much the information we need right? Why dont we store that one in the `UserDataManager`?

@Tillerino how serializable is that Settings object?

**Comment by ChemiHatake** [comment]
@omkelderman The Settings object contains settings for the recommendations

**Comment by omkelderman** [comment]
@ChemiHatake yeah, isnt that pretty much the information we're trying to store?

```
public static class Settings {
    public boolean nomod;
    public Model model;
    @BitwiseMods
    public long requestedMods;
    public List<RecommendationPredicate> predicates = new ArrayList<>();
}
```

Looks good to me :smile:

**Comment by ChemiHatake** [comment]
@omkelderman 
rip lazy me

ill work on it

**Comment by Tillerino** [comment]
I wouldn't serialize the `Settings` (or any similar) class.

By far the easiest and most flexible solution is having the default settings be a String and verifying that String with `tillerino.tillerinobot.RecommendationsManager.parseSamplerSettings(OsuApiUser, String, Language)`.

* If the settings space changes, the String might be invalid, but it will output a sensible error message. If a `Settings` (or similar) object is serialized and the class changes, deserialization might fail with unpredictable error messages.
* Verification should not be duplicated, any Settings object should not be duplicated.

But I have to say, I'd welcome more sophisticated checking of these rules in `tillerino.tillerinobot.RecommendationsManager.parseSamplerSettings(OsuApiUser, String, Language)`. Currently it is possible to go `!r beta dt` and AFAIK, `dt` will simply be ignored without warning.

**Comment by ChemiHatake** [comment]
@Tillerino Okay, so Ill add a new error message.
does that mean I'll have to add it in all the languages?

**Comment by ChemiHatake** [comment]
HOLY SHIT THIS TURNED OUT SO GUD

**Comment by ChemiHatake** [comment]
I'm gonna kill myself
fuq u hebrew

**Comment by ChemiHatake** [comment]
y u no build

**Comment by ChemiHatake** [comment]
for some reason it was searching for [nomod] when throwing an error in the Tsundere class.

**Comment by ChemiHatake** [comment]
14 commits wtf

File: tillerinobot/src/main/java/tillerino/tillerinobot/RecommendationsManager.java [hunk 2]
```diff
@@ -343,122 +342,137 @@ private GivenRecommendation toGivenRecommendation(
...
+		    
+		    
+		    for (int i = 0; i < remaining.length; i++) {
+
+		        String param = remaining[i];
+		        String lowerCase = param.toLowerCase();
+		        if (lowerCase.length() == 0) 
+		            continue;
+
+		        if (lowerCase.equals("any")) {
```
Comment (omkelderman) [inline tillerinobot/src/main/java/tillerino/tillerinobot/RecommendationsManager.java]: no `getLevenshteinDistance`? :stuck_out_tongue: seems to be used on "nomod"/"relax" and stuff... idk xd

**Comment by omkelderman** [comment]
> 14 commits wtf

Knowing tillerino they'll all be squashed into one commit anyway on merge :stuck_out_tongue: 

And about the languages, usally when adding a new message to the lang-interface, in every implementation where you do not know the language of you do something like this:

```
@Override
public String fooMethod() {
    return new Default().fooMethod();
}
```

(And i'll have to update the languages-wiki to set every lang to not-complete)

Maybe i'll rewrite the lang-structure such that everything extends from Default instead of implementing the Language-interface, cuz imo this is ugly xd, but imo its uglier to copy-pase the text itself :smile: 

Anyways, @Tillerino i got no further commends on this, too much has changed for me to comfortably say something about this, it is your code after all xd, so I'm gonna leave this up to you.

File: tillerinobot/src/main/java/tillerino/tillerinobot/UserDataManager.java [hunk 3]
```diff
@@ -205,6 +205,14 @@ public void setShowWelcomeMessage(boolean welcomeMessage) {
 			changed |= welcomeMessage != this.showWelcomeMessage;
 			this.showWelcomeMessage = welcomeMessage;
 		}
+		
+		@Getter
+		String recommendationDefault = "Gamma | Any.";
+		
+		public void setRecommendMods(String defaults) {
```
Comment (omkelderman) [inline tillerinobot/src/main/java/tillerino/tillerinobot/UserDataManager.java]: no `setRecommendationDefault`?

**Comment by ChemiHatake** [comment]
@omkelderman 
> no getLevenshteinDistance?
Any is a really short word, I don't think its necessary, 

> no setRecommendationDefault?
Doesn't really matter, but Ill change it in a sec

**Comment by omkelderman** [comment]
> Any is a really short word, is it necessary?

Idk, I just went with consistency cuz why not xd, dont really have a real reason :stuck_out_tongue:

**Comment by ChemiHatake** [comment]
@omkelderman dt hd and hr are small words and we use .equals for them heh

**Comment by MrMorkat** [comment]
Is this still being worked on? If yes, i can work on a solution for the tsundere conflicts.

**Comment by Tillerino** [comment]
Okay so basically what happened here is that the change became too complex
for the feature that it provided and we decided against merging it right
off the bat. @omk can you chime in here?

**Comment by omkelderman** [comment]
@Tillerino if you wanna mention me in the future, I recommend using my full username instead :wink: 

And I've kinda lost track of this PR tbh... Too much changes about stuff I don't know enough about (like those different models n shit), so I dont really know what to do.

**Pull request closed without merging**